chain-abci/src/liveness.rs

```diff
@@ -0,0 +1,102 @@
+use bit_vec::BitVec;
+use parity_scale_codec::{Decode, Encode, Error, Input, Output};
+use serde::{Deserialize, Serialize};
+
+use chain_core::state::account::StakedStateAddress;
+use chain_core::state::tendermint::BlockHeight;
+
+/// Liveness tracker for a validator
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct LivenessTracker {
+    /// Address of staking account
+    address: StakedStateAddress,
+    /// Holds data to measure liveness
+    liveness: BitVec,
+}
+
+impl LivenessTracker {
+    /// Creates a new instance of liveness tracker
+    #[inline]
+    pub fn new(address: StakedStateAddress, block_signing_window: u16) -> Self {
+        Self {
+            address,
+            liveness: BitVec::from_elem(block_signing_window as usize, true),
+        }
+    }
+
+    /// Updates liveness tracker with new block data
+    pub fn update(&mut self, block_height: BlockHeight, signed: bool) {
+        let block_signing_window = self.liveness.len();
+        let update_index = (block_height as usize - 1) % block_signing_window; // Because `block_height` starts from 1
+        self.liveness.set(update_index, signed)
+    }
+
+    /// Checks if validator is live or not
+    #[inline]
+    #[allow(dead_code)]
+    pub fn is_live(&self, missed_block_threshold: u16) -> bool {
+        let zero_count = self.liveness.iter().filter(|x| !x).count();
```

Comment:
ideally, this could be a single instruction `popcnt`, but this bit-vec implementation may not support it

Reply:
Yes. `bit-vec` crate does not provide this function.